### PR TITLE
Add remove null visitor

### PR DESCRIPTION
**Pull request by adalessa**

### CODE REVIEW

File: Tests/Service/Request/Visitor/RemoveNullFieldsRequestVisitorTest.php [hunk 1]
```diff
@@ -0,0 +1,135 @@
...
+
+        $this->uriInterfaceProfecy->__call('getPath', [])
+            ->willReturn("/v2/vehicles/43");
+
+        $request = $this->requestProphecy->reveal();
+        $factory = $this->streamFactoryProfecy->reveal();
+        $removeNullFieldsRequestVisitor = new RemoveNullFieldsRequestVisitor(
+            $factory,
+            [
+                new EndpointConfiguration("PUT", "/v2/vehicles/{id}")
```
Comment (aviator-ua) [inline Tests/Service/Request/Visitor/RemoveNullFieldsRequestVisitorTest.php:129]: Public repository. Do not expose internal endpoints

File: Service/Request/Visitor/EndpointConfiguration.php [hunk 2]
```diff
@@ -0,0 +1,58 @@
+<?php
+declare(strict_types=1);
```
Comment (aviator-ua) [inline Service/Request/Visitor/EndpointConfiguration.php:2]: I think, here CS warning — here should be one empty line

File: Service/Request/Visitor/EndpointConfiguration.php [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+
+namespace Auto1\ServiceAPIClientBundle\Service\Request\Visitor;
+
+/**
+ * Class EndpointConfiguration.
+ */
+class EndpointConfiguration
```
Comment (aviator-ua) [inline Service/Request/Visitor/EndpointConfiguration.php:9]: Should it be `EndpointConfiguration`, since we have already the same class in this package.
It could be quite confusing.

File: Service/Request/Visitor/RemoveNullFieldsRequestVisitor.php [hunk 4]
```diff
@@ -0,0 +1,96 @@
...
+    }
+
+    /**
+     * @return bool
+     */
+    private function isCompatible(RequestInterface $request): bool
+    {
+        foreach ($this->endpoints as $endpoint) {
+            if (
+                $request->getMethod() === $endpoint->getMethod()
```
Comment (aviator-ua) [inline Service/Request/Visitor/RemoveNullFieldsRequestVisitor.php:58]: I think, you should also check that you will process `json` stream.

File: Service/Request/Visitor/RemoveNullFieldsRequestVisitor.php [hunk 5]
```diff
@@ -0,0 +1,96 @@
...
+     */
+    private function clean($input): array
+    {
+        foreach ($input as &$value) {
+            if (is_array($value)) {
+                $value = $this->clean($value);
+            }
+        }
+
+        return array_filter($input, static function ($value) {
```
Comment (aviator-ua) [inline Service/Request/Visitor/RemoveNullFieldsRequestVisitor.php:92]: `array_filter` preserves array keys — so it means you will have `{"0" => "value", "2" => "value"}` after serialization

File: Tests/Service/Request/EndpointConfigurationTest.php [hunk 6]
```diff
@@ -0,0 +1,33 @@
...
+class EndpointConfigurationTest extends TestCase
+{
+    /**
+     * @dataProvider provider
+     */
+    public function testRegexGeneration($path, $regex)
+    {
+        $cut = new EndpointConfiguration("POST", $path);
+
+        $this->assertEquals($regex, $cut->getPathRegexp());
```
Comment (aviator-ua) [inline Tests/Service/Request/EndpointConfigurationTest.php:21]: In this package, assertions called as static function.

File: Tests/Service/Request/EndpointConfigurationTest.php [hunk 7]
```diff
@@ -0,0 +1,33 @@
...
+ * Class EndpointConfigurationTest.
+ */
+class EndpointConfigurationTest extends TestCase
+{
+    /**
+     * @dataProvider provider
+     */
+    public function testRegexGeneration($path, $regex)
+    {
+        $cut = new EndpointConfiguration("POST", $path);
```
Comment (aviator-ua) [inline Tests/Service/Request/EndpointConfigurationTest.php:19]: Pls, use singlequotes

File: Tests/Service/Request/Visitor/RemoveNullFieldsRequestVisitorTest.php [hunk 8]
```diff
@@ -0,0 +1,135 @@
...
+            ]
+        );
+
+        $removeNullFieldsRequestVisitor->visit($request);
+    }
+
+    public function provider()
+    {
+        return [
```
Comment (aviator-ua) [inline Tests/Service/Request/Visitor/RemoveNullFieldsRequestVisitorTest.php:91]: What will be if all elements will be removed?

File: Tests/Service/Request/Visitor/RemoveNullFieldsRequestVisitorTest.php [hunk 9]
```diff
@@ -0,0 +1,135 @@
...
+            'remove first level null' => [
+                json_encode(["a" => 1, "b" => null]),
+                json_encode(["a" => 1])
+            ], 'remove sub level null' => [
```
Comment (aviator-ua) [inline Tests/Service/Request/Visitor/RemoveNullFieldsRequestVisitorTest.php:95]: Formating

**Comment by aviator-ua** [comment]
Just like a suggestion, could it be part of the "serialization" process?
For me, this is not looking good when the client modifies the request.

File: Tests/Service/Request/Visitor/RemoveNullFieldsRequestVisitorTest.php (hunk 1 above)
Comment (adalessa) [reply Tests/Service/Request/Visitor/RemoveNullFieldsRequestVisitorTest.php]: Done, changed to generic

File: Service/Request/Visitor/EndpointConfiguration.php (hunk 2 above)
Comment (adalessa) [reply Service/Request/Visitor/EndpointConfiguration.php:3]: New line added

File: Tests/Service/Request/EndpointConfigurationTest.php (hunk 7 above)
Comment (adalessa) [reply Tests/Service/Request/EndpointConfigurationTest.php]: changed

File: Tests/Service/Request/Visitor/RemoveNullFieldsRequestVisitorTest.php (hunk 9 above)
Comment (adalessa) [reply Tests/Service/Request/Visitor/RemoveNullFieldsRequestVisitorTest.php]: fix

File: Tests/Service/Request/Visitor/RemoveNullFieldsRequestVisitorTest.php (hunk 8 above)
Comment (adalessa) [reply Tests/Service/Request/Visitor/RemoveNullFieldsRequestVisitorTest.php:92]: In case of all the elements are deleted the parent element will remain empty.

Comment (aviator-ua) [reply Tests/Service/Request/Visitor/RemoveNullFieldsRequestVisitorTest.php:92]: And this empty element will be encoded into `{}` or `[]`? 
It's a quite important question for none-PHP services

File: Service/Request/Visitor/RemoveNullFieldsRequestVisitor.php (hunk 5 above)
Comment (adalessa) [reply Service/Request/Visitor/RemoveNullFieldsRequestVisitor.php:93]: I see the point, from the original ticket there is no a clear answer what to do in this case. I don't know if we are counting for cases of just array of elements without key to be null. Generally arrays of elements are object not just values.

File: Service/Request/Visitor/EndpointConfiguration.php [hunk 10]
```diff
@@ -0,0 +1,59 @@
...
+    {
+        return $this->path;
+    }
+
+    /**
+     * @return string
+     */
+    public function getPathRegexp(): string
+    {
+        $regex = preg_replace('/{(\w*)}/', '\w+', $this->getPath());
```
Comment (aviator-ua) [inline Service/Request/Visitor/EndpointConfiguration.php:55]: You should "escape" also all regex special symbols, that are valid for URI, for example `?`

File: Tests/Service/Request/Visitor/RemoveNullFieldsRequestVisitorTest.php (hunk 8 above)
Comment (adalessa) [reply Tests/Service/Request/Visitor/RemoveNullFieldsRequestVisitorTest.php:92]: Will be encoded as `[]`

File: Service/Request/Visitor/EndpointConfiguration.php (hunk 10 above)
Comment (adalessa) [reply Service/Request/Visitor/EndpointConfiguration.php:55]: Added escape for question mark

File: Tests/Service/Request/Visitor/RemoveNullFieldsRequestVisitorTest.php (hunk 8 above)
Comment (aviator-ua) [reply Tests/Service/Request/Visitor/RemoveNullFieldsRequestVisitorTest.php:92]: Same question about empty assoc array. 

Could you cover these corner cases with tests also?

File: Service/Request/Visitor/RemoveNullFieldsRequestVisitor.php (hunk 4 above)
Comment (adalessa) [reply Service/Request/Visitor/RemoveNullFieldsRequestVisitor.php:59]: And if the is not a json, should it throw an exception?

File: Service/Request/Visitor/RemoveNullFieldsRequestVisitor.php [hunk 11]
```diff
@@ -0,0 +1,97 @@
...
+
+        return false;
+    }
+
+    /**
+     * @return StreamInterface
+     */
+    private function cleanBody(StreamInterface $requestBody): StreamInterface
+    {
+        $data = json_decode($requestBody->getContents(), true);
```
Comment (aivus) [inline Service/Request/Visitor/RemoveNullFieldsRequestVisitor.php:74]: I'm not sure that we can do `json_decode`/`json_encode` here as instead of using `auto1.api.request.serializer` as the original application does.

It's important to encode/decode in the same way using the same serializer(with same configuration) as original query.

File: Service/Request/Visitor/RemoveNullFieldsRequestVisitor.php (hunk 5 above)
Comment (aviator-ua) [reply Service/Request/Visitor/RemoveNullFieldsRequestVisitor.php:93]: Since it is `auto1-oss`, we talk not only about internal usage. So, for sure, it could be a simple array of values. 
After you will remove null-elements, it still should be an array (I mean json-representation, not PHP object).
